sdk/samples/iot/iot_sample_foundation.c

```diff
@@ -0,0 +1,450 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "iot_sample_foundation.h"
+

```

Comment:
why? I never saw those included in other files to my knowledge, and everything compiles and samples run.

Reply:
So it'll compile okay because the necessary includes are in the header. However, generally files should be able to exist on their own in terms of "dependencies". So the `.c` file should include everything needed to work in its include section. Any structs it references in function declarations, implementation, etc should have an according include if needed. The `.c` shouldn't rely on its according header (in this case `iot_sample_foundation.h`) to include everything for it.

Reply:
Interesting. I am wondering about the pros/cons of repeating the .h includes in the .c file to be explicit, vs relying on the includes in the header file. Maybe, there is a middle ground where foo.c doesn't need to repeat the headers already included in foo.h, since foo.c will certainly have to include foo.h anyway.

And then there is a question of transitive includes.

For example, `az_span_internal.h` has to include the public `az_span.h`, so if I have a .c file that include `az_span_internal.h`, should I also include `az_span.h`? From an "access" perspective, of course if I have access to internals, it could be a given that I have access to the public header too.

Reply:
good question @ahsonkhan, @danewalton and I were just talking about this.  He helped me see the value of being explicit with the needs, so it would be valuable to include `az_span_internal.h` and `az_span.h` only if there are elements within `az_span.h` that need to be accessed that `az_span_internal.h` does not provide.  We were discussing the similar scenario of `stddef.h` vs `stdlib.h`.  

I have worked on a project were this was not explicit, and as a developer one has no idea how certain headers are being included because they are so deeply buried in the chain of #includes.  It can be quite confusing.  I agree with Dane that a good approach is to scope the include as small as possible and only include the "superset" include file if something needs to be accessed outside of the initial set.  May be a good topic for next C99 meeting :)